wazimap_ng/datasets/views.py

```diff
@@ -20,15 +23,88 @@
 
-class DatasetList(generics.ListAPIView):
+class DatasetList(generics.ListCreateAPIView):
     queryset = models.Dataset.objects.all()
     serializer_class = serializers.DatasetSerializer
 
+
+    def post(self, request, *args, **kwargs):
+        profile_id = request.data.get("profile")
+        profile = get_object_or_404(Profile, pk=profile_id)
+        if not request.user.groups.filter(name=profile.name).exists():
+            return Response({
+                    'detail': 'You do not have permission to upload to this profile'
+                }, status=status.HTTP_403_FORBIDDEN
+            )
+        response = self.create(request, *args, **kwargs)
+
+        dataset_obj = models.Dataset.objects.filter(id=response.data["id"]).first()
+
+        # Assign dataset object permission to profile group
+        assign_perms_to_group(profile.name, dataset_obj, False)
+
+        # Get and create dataset file if user has sent file while creating dataset
+        file_obj = request.data.get('file', None)
+        if file_obj:
+            datasetfile_obj = models.DatasetFile.objects.create(
+                name=dataset_obj.name,
+                document=file_obj,
+                dataset_id=dataset_obj.id
+            )
+
+            task = async_task(
+                "wazimap_ng.datasets.tasks.process_uploaded_file",
+                datasetfile_obj, dataset_obj,
+                task_name=f"Uploading data: {dataset_obj.name}",
+                hook="wazimap_ng.datasets.hooks.process_task_info",
+                key=request.session.session_key,
+                type="upload", assign=True, notify=True
+            )
+
+        return response
+
 class DatasetDetailView(generics.RetrieveAPIView):
     queryset = models.Dataset
     serializer_class = serializers.DatasetDetailViewSerializer
 
+
+@api_view(['POST'])
+def dataset_upload(request, dataset_id):
+    dataset = get_object_or_404(models.Dataset, pk=dataset_id)
+
+    if not request.user.groups.filter(name=dataset.profile.name).exists():
+        return Response({
+                'detail': 'You do not have permission to upload to this profile'
+            }, status=status.HTTP_403_FORBIDDEN
+        )
+    file_obj = request.data.get('file', None)
+
+    if not file_obj:
+        return Response({
+                'detail': 'please add file for upload'
+            }, status=status.HTTP_403_FORBIDDEN
+        )
+
+    datasetfile_obj = models.DatasetFile.objects.create(
+        name=dataset.name,
+        document=file_obj,
+        dataset_id=dataset.id
+    )
+
+    task = async_task(
+        "wazimap_ng.datasets.tasks.process_uploaded_file",
+        datasetfile_obj, dataset,
+        task_name=f"Uploading data: {dataset.name}",
+        hook="wazimap_ng.datasets.hooks.process_task_info",
+        key=request.session.session_key,
+        type="upload", assign=True, notify=True
+    )
+
+    return Response({'detail': 'uploaded dataset'}, status=status.HTTP_200_OK)
```

Comment:
let's return the task id as well. 
so that we can create another endpoint for polling the status of that task.